tls/s2n_connection_handles.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2017 Amazon.com = Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License = Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS = WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND = either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_connection_handles.h"
+
+#include "utils/s2n_safety.h"
+
+int s2n_connection_save_prf_state(struct s2n_connection_prf_handles *prf_handles, struct s2n_connection *conn)
+{
+    /* Preserve handlers for TLS PRF p_hash to avoid re-allocation */
+    prf_handles->p_hash_s2n_hmac_inner = conn->prf_space.tls.p_hash.s2n_hmac.inner.digest.evp;
+    prf_handles->p_hash_s2n_hmac_inner_impl = conn->prf_space.tls.p_hash.s2n_hmac.inner.hash_impl;
+    prf_handles->p_hash_s2n_hmac_inner_just_key = conn->prf_space.tls.p_hash.s2n_hmac.inner_just_key.digest.evp;
+    prf_handles->p_hash_s2n_hmac_inner_just_key_impl = conn->prf_space.tls.p_hash.s2n_hmac.inner_just_key.hash_impl;
+    prf_handles->p_hash_s2n_hmac_outer = conn->prf_space.tls.p_hash.s2n_hmac.outer.digest.evp;
+    prf_handles->p_hash_s2n_hmac_outer_impl = conn->prf_space.tls.p_hash.s2n_hmac.outer.hash_impl;
+    prf_handles->p_hash_evp_hmac = conn->prf_space.tls.p_hash.evp_hmac;
+    prf_handles->p_hash_hmac = conn->prf_space.tls.p_hash_hmac;
+
+    return 0;
+}
+
+int s2n_connection_save_hash_state(struct s2n_connection_hash_handles *hash_handles, struct s2n_connection *conn)
+{
+    /* Preserve handlers for handshake hash states to avoid re-allocation */
+    hash_handles->md5 = conn->handshake.md5.digest.evp;
+    hash_handles->md5_hash_impl = conn->handshake.md5.hash_impl;
```

Comment:
I have pulled in changes from what was going to be another PR (#547) to add global state on which hash implementation to use rather than this per-hash state. That change eliminates the need to preserve any hash_impl's on s2n_connection_wipe.